From e0ea1489ee129cd65b62f41820b98ea8b1e25e29 Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Mon, 2 Mar 2015 14:09:29 +0300
Subject: [PATCH 09/N] convert(): be more careful to stay within the buffer

We really must not run outside of the buffer. And the buffer needs to be
NUL-terminated, too...

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 winsup/cygwin/msys2_path_conv.cc | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/winsup/cygwin/msys2_path_conv.cc b/winsup/cygwin/msys2_path_conv.cc
index 4fdcda2..56b638e 100644
--- a/winsup/cygwin/msys2_path_conv.cc
+++ b/winsup/cygwin/msys2_path_conv.cc
@@ -523,7 +523,7 @@ void rp_convert(const char** from, const char* to, char** dst, const char* dsten
         posix_to_win32_path(it, real_to, dst, dstend);
     }
 
-    if (real_to != to) {
+    if (*dst != dstend && real_to != to) {
         **dst = *real_to;
         *dst += 1;
     }
@@ -579,6 +579,9 @@ void ppl_convert(const char** from, const char* to, char** dst, const char* dste
             subp_convert(&beg, it, is_url, dst, dstend);
             is_url = 0;
 
+            if (*dst == dstend)
+	        break;
+
             **dst = ';';
             *dst += 1;
         }
-- 
2.9.0

